File: trainer/domain-config.mjs
// The Forge — Domain Configuration Loader
// Reads domain-specific config from domains/*.json
// All pipeline scripts use this to stay domain-agnostic.
//
// Usage:
//   import { loadDomain, getActiveDomain } from "./domain-config.mjs";
//   const domain = loadDomain("legal");  // or getActiveDomain() for current

import { readFileSync, readdirSync, existsSync, writeFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DOMAINS_DIR = resolve(__dirname, "domains");
const ACTIVE_FILE = resolve(__dirname, "domains", ".active");

const DEFAULT_DOMAIN = "coding";

/** Load a domain config by ID */
export function loadDomain(domainId) {
  const file = resolve(DOMAINS_DIR, `${domainId}.json`);
  if (!existsSync(file)) {
    throw new Error(`Domain "${domainId}" not found at ${file}`);
  }
  return JSON.parse(readFileSync(file, "utf8"));
}

/** Get the currently active domain ID */
export function getActiveDomainId() {
  if (existsSync(ACTIVE_FILE)) {
    return readFileSync(ACTIVE_FILE, "utf8").trim() || DEFAULT_DOMAIN;
  }
  return DEFAULT_DOMAIN;
}

/** Get the currently active domain config */
export function getActiveDomain() {
  return loadDomain(getActiveDomainId());
}

/** Set the active domain */
export function setActiveDomain(domainId) {
  // Validate it exists
  loadDomain(domainId);
  writeFileSync(ACTIVE_FILE, domainId + "\n");
}

/** List all available domains */
export function listDomains() {
  const files = readdirSync(DOMAINS_DIR).filter(
    (f) => f.endsWith(".json") && !f.startsWith(".")
  );
  return files.map((f) => {
    const domain = JSON.parse(readFileSync(resolve(DOMAINS_DIR, f), "utf8"));
    return {
      id: domain.id,
      name: domain.name,
      description: domain.description,
      model_prefix: domain.model_prefix,
      active: domain.id === getActiveDomainId(),
    };
  });
}

/** Get the Ollama URL for a domain — remote GPU if configured, else local */
export function getOllamaUrl(domain) {
  // Env override always wins
  if (process.env.OLLAMA_URL) return process.env.OLLAMA_URL;

  // Use remote GPU endpoint if domain has one configured
  if (domain?.remote?.ollama_url) return domain.remote.ollama_url;

  // Fallback to local
  return "http://localhost:11434";
}

export default { loadDomain, getActiveDomain, getActiveDomainId, setActiveDomain, listDomains, getOllamaUrl };

File: trainer/scripts/domain_config.py
"""
The Forge — Domain Configuration (Python side)
Reads domain-specific config from domains/*.json
Shared by prepare-data.py, evaluate.py, train.py, and fuse-and-deploy.py
"""

import json
import os
from pathlib import Path

TRAINER_DIR = Path(__file__).resolve().parent.parent
DOMAINS_DIR = TRAINER_DIR / "domains"
ACTIVE_FILE = DOMAINS_DIR / ".active"
DEFAULT_DOMAIN = "coding"


def load_domain(domain_id):
    """Load a domain config by ID."""
    f = DOMAINS_DIR / f"{domain_id}.json"
    if not f.exists():
        raise FileNotFoundError(f'Domain "{domain_id}" not found at {f}')
    return json.loads(f.read_text())


def get_active_domain_id():
    """Get the currently active domain ID."""
    if ACTIVE_FILE.exists():
        text = ACTIVE_FILE.read_text().strip()
        return text if text else DEFAULT_DOMAIN
    return DEFAULT_DOMAIN


def get_active_domain():
    """Get the currently active domain config."""
    return load_domain(get_active_domain_id())


def set_active_domain(domain_id):
    """Set the active domain."""
    load_domain(domain_id)  # validate exists
    ACTIVE_FILE.write_text(domain_id + "\n")


def list_domains():
    """List all available domains."""
    active = get_active_domain_id()
    domains = []
    for f in sorted(DOMAINS_DIR.glob("*.json")):
        if f.name.startswith("."):
            continue
        d = json.loads(f.read_text())
        domains.append({
            "id": d["id"],
            "name": d["name"],
            "description": d["description"],
            "model_prefix": d["model_prefix"],
            "active": d["id"] == active,
        })
    return domains
